### PR TITLE
fix(migrate)!: app router and next-intl

### LINKED ISSUE #61: migrate to app router

### PR DESCRIPTION
**Pull request by GMSoudersJr**
I keep getting what looks like the same error for all pages except home, but the messages do not help. There is one more thing I will try but I wanted to get some advice before I spend more time fixing this.

BREAKING CHANGE

### CODE REVIEW

**Comment by GMSoudersJr** [comment]
I have been going over this for days, but am stuck. I have tried adding a linter for hooks, comparing versions of react and react-dom, moving everything into the app router, I don't know else to do. It is probably something simple that I am not seeing. Yay React!

**Comment by ivanms1** [comment]
thanks for taking the time to work on this, I'll take a look today

**Comment by GMSoudersJr** [comment]
No rush at all. One plus of working on this is that it got me back into running. I had to take time to clear my head after some of this.

**Comment by ivanms1** [comment]
I think the problem is the NextSeo component, after removing it the error went away.
Probably not compatible with app router, yay
On the other hand you could use the now built in [metada api](https://nextjs.org/docs/app/building-your-application/optimizing/metadata)
hopefully that helps

**Comment by GMSoudersJr** [comment]
Okay, that figures. I did notice that the title wasn't visible in the browser tab. New inspiration!

**Comment by GMSoudersJr** [comment]
I apologize for how large this is.

Your suggestion worked like a charm. Those errors are gone and it builds successfully.

I implemented the built-in metadata from next.js. The site seems to have the same functionality it did before. Sharing the link needs to be tested, though.

File: .eslintrc [hunk 1]
```diff
@@ -5,7 +5,7 @@
     "node": true
   },
   "extends": ["next", "next/core-web-vitals", "prettier"],
-  "plugins": ["@typescript-eslint"],
+  "plugins": ["@typescript-eslint", "react-hooks"],
   "parser": "@typescript-eslint/parser",
   "settings": {
     "react": {
```
Comment (ivanms1) [inline .eslintrc]: could you revert this?
I think this plugin is already included in the next plugin

File: .eslintrc [hunk 2]
```diff
@@ -18,6 +18,8 @@
     "@typescript-eslint/explicit-module-boundary-types": "off",
     "@typescript-eslint/camelcase": "off",
     "@typescript-eslint/no-explicit-any": "off",
-    "@typescript-eslint/no-unused-vars": ["error"]
+    "@typescript-eslint/no-unused-vars": ["error"],
+    "react-hooks/rules-of-hooks": "error",
+    "react-hooks/exhaustive-deps": "warn"
```
Comment (ivanms1) [inline .eslintrc]: these also can go I think

File: src/app/[locale]/components/Button/Button.tsx [hunk 3]
```diff
@@ -1,3 +1,5 @@
+"use client";
```
Comment (ivanms1) [inline src/app/[locale]/components/Button/Button.tsx]: is it possible to move the `components` folder outside of the `app` folder.

File: src/app/[locale]/components/Layout/Layout.tsx [hunk 4]
```diff
@@ -1,7 +1,9 @@
-import React from "react";
+"use client";
 
-import Footer from "./Footer";
-import Navbar from "./Navbar";
+//import React from "react";
```
Comment (ivanms1) [inline src/app/[locale]/components/Layout/Layout.tsx]: please remove

File: src/app/[locale]/components/Layout/Navbar/Navbar.tsx [hunk 5]
```diff
@@ -1,22 +1,24 @@
...
-import { useTranslation } from "next-i18next";
-
+//import React from "react";
+import { Link, useRouter, usePathname } from "src/navigation";
+import { useTranslations } from "next-intl";
 import styles from "./Navbar.module.scss";
-import { useRouter } from "next/router";
-import Button from "@/components/Button";
+import { useParams } from "next/navigation";
+import Button from "../../Button/Button";
```
Comment (ivanms1) [inline src/app/[locale]/components/Layout/Navbar/Navbar.tsx]: always prefer absolute imports

File: src/app/[locale]/components/Layout/Navbar/Navbar.tsx [hunk 6]
```diff
@@ -1,22 +1,24 @@
...
 
   const handleLocaleChange = () => {
     if (locale === LOCALES.ko.code) {
-      push({ pathname, query }, asPath, { locale: LOCALES.en.code });
+      push(`${pathname}/`, { locale: "en" });
       return;
     }
 
-    push({ pathname, query }, asPath, { locale: LOCALES.ko.code });
+    push(`${LOCALES.ko.code}/${pathname}/`, { locale: "ko" });
```
Comment (ivanms1) [inline src/app/[locale]/components/Layout/Navbar/Navbar.tsx]:
```suggestion
    push(`${LOCALES.ko.code}/${pathname}/`, { locale: LOCALES.ko.code });
```

File: src/app/[locale]/components/MemberCard/MemberCard.tsx [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import React from "react";
+//import React from "react";
```
Comment (ivanms1) [inline src/app/[locale]/components/MemberCard/MemberCard.tsx]: remove please

File: src/app/[locale]/events/Events.tsx [hunk 8]
```diff
@@ -1,21 +1,21 @@
...
+
+import meetupImage from "@/assets/images/fcc-location.jpg";
+import kakaoMapsLogo from "@/assets/images/kakao-maps.png";
+import naverMapsLogo from "@/assets/images/naver-maps.png";
+
 const NAVER_MAPS_URL = "https://naver.me/xaC9isrW";
 const KAKAO_MAPS_URL = "http://kko.to/fkhQtFpFz";
 
-interface EventsProps {}
+//interface EventsProps {}
```
Comment (ivanms1) [inline src/app/[locale]/events/Events.tsx]: this was here from before, please remove if not needed

File: src/app/[locale]/events/page.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+export const metadata: Metadata = {
+  title: "Events",
+};
+
+type Props = {
+  params: { locale: string };
+};
+
+export default async function Page({ params: { locale } }: Props) {
+  unstable_setRequestLocale(locale);
```
Comment (ivanms1) [inline src/app/[locale]/events/page.tsx:14]: what a name 😆

File: src/app/[locale]/layout.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  return locales.map((locale) => ({ locale }));
+}
+
+export default async function LocaleLayout({
+  children,
+  params: { locale },
+}: Props) {
+  unstable_setRequestLocale(locale);
+  // Providing all messages to the client
+  // is the easiest way to get started
```
Comment (ivanms1) [inline src/app/[locale]/layout.tsx]: are these comments from the docs?

File: src/app/layout.tsx [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import { ReactNode } from "react";
+import "./globals.scss";
+
+type Props = {
+  children: ReactNode;
+};
+
+export default function RootLayout({ children }: Props) {
```
Comment (ivanms1) [inline src/app/layout.tsx:8]: we could move `src/app/[locale]/components/Layout/Layout.tsx` here, but keep the Navbar and Footer in the components folder.

File: src/i18n.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { notFound } from "next/navigation";
+import { getRequestConfig } from "next-intl/server";
+
+// Can be import from a shared config
+const locales = ["en", "ko"];
```
Comment (ivanms1) [inline src/i18n.ts]: maybe we could move this to `src/const/index.ts`
```suggestion
const LOCALES = ["en", "ko"];
```

File: src/services/members/index.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import { MEMBERS_API } from "@/const";
+import { MEMBERS_API, revalidateTime } from "@/const";
```
Comment (ivanms1) [inline src/services/members/index.ts]: could you rename this to `REVALIDATE_TIME`?

File: src/services/meetup/index.ts [hunk 14]
```diff
@@ -1,7 +1,7 @@
 import { MEETUP_LINK } from "@/const";
 
 export const getMeetupInfo = async () => {
-  const meetupResponse = await fetch(MEETUP_LINK);
+  const meetupResponse = await fetch(MEETUP_LINK, { cache: "no-store" });
```
Comment (ivanms1) [inline src/services/meetup/index.ts:4]: `{ cache: "no-store" }`
this makes it server side rendered?

**Review by ivanms1 — COMMENTED** [review]
Great job!
thank you for the hard work, left some minor comments.

File: src/app/[locale]/layout.tsx (hunk 10 above)
Comment (GMSoudersJr) [reply src/app/[locale]/layout.tsx]: They are. I left them for the next person who contributes. I'll remove them, though.

File: src/services/meetup/index.ts (hunk 14 above)
Comment (GMSoudersJr) [reply src/services/meetup/index.ts:4]: Correct. This is the app router way of calling getServerSideProps by using options in fetch. [Docs](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration#step-6-migrating-data-fetching-methods)

File: .eslintrc (hunk 1 above)
Comment (GMSoudersJr) [reply .eslintrc]: should I `yarn remove <package-name>` since I added it with yarn? Or simply delete it from the .eslintrc file?

File: src/app/layout.tsx (hunk 11 above)
Comment (GMSoudersJr) [reply src/app/layout.tsx:8]: 🤔 hmmm... Okay. Hadn't crossed my mind.

File: .eslintrc (hunk 1 above)
Comment (ivanms1) [reply .eslintrc]: both please 🙇

File: src/services/meetup/index.ts (hunk 14 above)
Comment (ivanms1) [reply src/services/meetup/index.ts:4]: neat 👍

File: src/app/layout.tsx (hunk 11 above)
Comment (GMSoudersJr) [reply src/app/layout.tsx:8]: I am a little confused by this.

I used this structure to catch errors (create an error boundary) because I was tired of getting those ugly error message windows when I could not get the locale to work. [next-intl docs](https://next-intl-docs.vercel.app/docs/environments/error-files).

Are you suggesting to move the `src/app/[locale]/compnents/Layout/Layout.tsx` file to `src/app` directory?
Would it not be better to place it in `src/app/[locale]/` directory since it is the layout that is used with a locale.

Comment (ivanms1) [reply src/app/layout.tsx:8]: I see what you mean now, this layout is not doing much but it seems that is required because the build fails when it's missing.
We can just leave it as it is, only one thing, could you move BaseLayout and its modules to the components folder?

**Comment by ivanms1** [comment]
looking good, just one more thing, I think we can uninstall next-seo

**Comment by GMSoudersJr** [comment]
it's weird. next-seo was not in my package.json

**Comment by ivanms1** [comment]
you removed it already 😄

File: src/components/MemberCard/MemberCard.tsx [hunk 15]
```diff
@@ -1,4 +1,4 @@
-import React from "react";
+//import React from "react";
```
Comment (ivanms1) [inline src/components/MemberCard/MemberCard.tsx]: last one!

File: tsconfig.json [hunk 16]
```diff
@@ -23,6 +30,6 @@
     },
     "incremental": true
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
```
Comment (ivanms1) [inline tsconfig.json:33]: could you tell me why was this needed?

File: src/services/members/index.ts [hunk 17]
```diff
@@ -9,9 +9,12 @@ export interface MemberData {
...
 
 export const getMemberInfo = async (): Promise<MemberData[]> => {
-  const res = await fetch(MEMBERS_API);
+  const res = await fetch(MEMBERS_API, {
+    next: { revalidate: REVALIDATE_TIME },
+  });
+
   const responseJSON = await res.json();
-  const memberData = responseJSON.data;
+  const memberData = await responseJSON.data;
```
Comment (ivanms1) [inline src/services/members/index.ts]: I don't think you need to await this

**Review by ivanms1 — COMMENTED** [review]
last few comments
almost there!

File: src/services/members/index.ts (hunk 17 above)
Comment (GMSoudersJr) [reply src/services/members/index.ts]: Correct

File: tsconfig.json (hunk 16 above)
Comment (GMSoudersJr) [reply tsconfig.json:33]: Honestly, I do not remember adding that. Maybe it was added during one of my yarn commands.

I looked into it more, and found that I built the app before pushing my commits. `yarn build` gave this message:
image

Comment (GMSoudersJr) [reply tsconfig.json:33]: I guess this is not a best practice since it is optimizing for production and not what we use for development.

Comment (ivanms1) [reply tsconfig.json:33]: oh that's interesting, I guess we can keep it then

**Review by ivanms1 — APPROVED** [review]
LGTM 🎉 
Thank you for taking the time to do this.

**Comment by GMSoudersJr** [comment]
Thanks for taking the time to review it.
